File: android/app/src/main/java/com/modi/connect/session/SessionControlMessage.kt
package com.modi.connect.session

import com.modi.protocol.LinkType
import com.modi.protocol.Packet
import com.modi.protocol.PacketType
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.charset.StandardCharsets
import java.util.UUID

enum class SessionControlAction(val code: Byte) {
    DISCONNECT_REQUEST(1),
    DISCONNECT_ACK(2);

    companion object {
        fun fromCode(code: Byte): SessionControlAction? = entries.firstOrNull { it.code == code }
    }
}

enum class DisconnectReason(val code: Byte) {
    USER_SWITCH(1),
    REPAIR(2),
    USER_STOP(3);

    companion object {
        fun fromCode(code: Byte): DisconnectReason? = entries.firstOrNull { it.code == code }
    }
}

enum class DisconnectResult(val code: Byte) {
    NONE(0),
    ACCEPTED(1),
    IGNORED(2);

    companion object {
        fun fromCode(code: Byte): DisconnectResult? = entries.firstOrNull { it.code == code }
    }
}

data class SessionControlMessage(
    val action: SessionControlAction,
    val sessionId: UUID,
    val oldLink: Byte,
    val targetLink: Byte,
    val reason: DisconnectReason,
    val result: DisconnectResult
) {
    fun encode(): ByteArray = ByteArray(MESSAGE_LENGTH).also { payload ->
        payload[0] = DOMAIN
        payload[1] = MESSAGE_TYPE
        payload[2] = VERSION
        payload[3] = action.code
        sessionId.writeNetworkBytes(payload, SESSION_OFFSET)
        payload[20] = oldLink
        payload[21] = targetLink
        payload[22] = reason.code
        payload[23] = result.code
    }

    fun toPacket(): Packet = Packet(PacketType.DATA, oldLink, 0u, encode())

    companion object {
        private const val MESSAGE_LENGTH = 24
        private const val SESSION_OFFSET = 4
        private const val DOMAIN: Byte = 0x0f
        private const val MESSAGE_TYPE: Byte = 0x01
        private const val VERSION: Byte = 0x01

        fun request(
            sessionId: UUID,
            oldLink: Byte,
            targetLink: Byte,
            reason: DisconnectReason
        ): SessionControlMessage {
            require(isLinkType(oldLink) && isLinkType(targetLink)) { "Unsupported link type" }
            return SessionControlMessage(
                SessionControlAction.DISCONNECT_REQUEST,
                sessionId,
                oldLink,
                targetLink,
                reason,
                DisconnectResult.NONE
            )
        }

        fun ack(request: SessionControlMessage, result: DisconnectResult): SessionControlMessage {
            require(request.action == SessionControlAction.DISCONNECT_REQUEST) { "ACK requires a request" }
            require(result != DisconnectResult.NONE) { "ACK result must be accepted or ignored" }
            return request.copy(action = SessionControlAction.DISCONNECT_ACK, result = result)
        }

        fun decode(payload: ByteArray): SessionControlMessage? {
            if (payload.size != MESSAGE_LENGTH ||
                payload[0] != DOMAIN || payload[1] != MESSAGE_TYPE || payload[2] != VERSION
            ) return null

            val action = SessionControlAction.fromCode(payload[3]) ?: return null
            val oldLink = payload[20]
            val targetLink = payload[21]
            if (!isLinkType(oldLink) || !isLinkType(targetLink)) return null
            val reason = DisconnectReason.fromCode(payload[22]) ?: return null
            val result = DisconnectResult.fromCode(payload[23]) ?: return null
            if (action == SessionControlAction.DISCONNECT_REQUEST && result != DisconnectResult.NONE) return null
            if (action == SessionControlAction.DISCONNECT_ACK && result == DisconnectResult.NONE) return null

            return SessionControlMessage(
                action,
                uuidFromNetworkBytes(payload, SESSION_OFFSET),
                oldLink,
                targetLink,
                reason,
                result
            )
        }
    }
}

data class HelloSessionIdentity(
    val route: Int,
    val token: String?,
    val sessionId: UUID
)

data class HandshakeResult(val route: Int, val sessionId: UUID)

data class P2pHandshakeResult(val remoteIp: String, val sessionId: UUID)

object HelloSessionPayload {
    private const val TOKEN_LENGTH = 8
    private const val SESSION_LENGTH = 16

    fun encode(route: Int, token: String?, sessionId: UUID): ByteArray {
        require(route in 0..3) { "Route must be between 0 and 3" }
        val tokenBytes = token?.toByteArray(StandardCharsets.US_ASCII)
        require(tokenBytes == null || tokenBytes.isNotEmpty() && tokenBytes.size <= TOKEN_LENGTH) {
            "Token must contain 1 to 8 ASCII bytes"
        }
        require(token == null || tokenBytes!!.toString(StandardCharsets.US_ASCII) == token) {
            "Token must contain ASCII characters only"
        }

        val sessionOffset = 1 + (if (tokenBytes == null) 0 else TOKEN_LENGTH)
        return ByteArray(sessionOffset + SESSION_LENGTH).also { payload ->
            payload[0] = route.toByte()
            tokenBytes?.copyInto(payload, destinationOffset = 1)
            sessionId.writeNetworkBytes(payload, sessionOffset)
        }
    }

    fun decode(payload: ByteArray, tokenRequired: Boolean): HelloSessionIdentity? {
        val sessionOffset = if (tokenRequired) 1 + TOKEN_LENGTH else 1
        if (payload.size != sessionOffset + SESSION_LENGTH) return null
        val route = payload[0].toInt()
        if (route !in 0..3) return null

        val token = if (tokenRequired) {
            val tokenEnd = (1 until sessionOffset).firstOrNull { payload[it] == 0.toByte() } ?: sessionOffset
            if ((tokenEnd until sessionOffset).any { payload[it] != 0.toByte() }) return null
            if (tokenEnd == 1) return null
            val decoded = payload.copyOfRange(1, tokenEnd).toString(StandardCharsets.US_ASCII)
            if (!decoded.all { it.code in 0x20..0x7e }) return null
            decoded
        } else {
            null
        }

        return HelloSessionIdentity(route, token, uuidFromNetworkBytes(payload, sessionOffset))
    }

    fun matchesAck(payload: ByteArray, expectedSessionId: UUID): Boolean =
        decode(payload, tokenRequired = false)?.sessionId == expectedSessionId
}

private fun isLinkType(linkType: Byte): Boolean = when (linkType) {
    LinkType.WIFI_LAN, LinkType.WIFI_DIRECT, LinkType.BLUETOOTH, LinkType.USB -> true
    else -> false
}

private fun UUID.writeNetworkBytes(destination: ByteArray, offset: Int) {
    ByteBuffer.wrap(destination, offset, 16)
        .order(ByteOrder.BIG_ENDIAN)
        .putLong(mostSignificantBits)
        .putLong(leastSignificantBits)
}

private fun uuidFromNetworkBytes(source: ByteArray, offset: Int): UUID {
    val buffer = ByteBuffer.wrap(source, offset, 16).order(ByteOrder.BIG_ENDIAN)
    return UUID(buffer.long, buffer.long)
}

File: windows/MoDi.Desktop/Core/Session/SessionControlMessage.cs
using System;
using System.Text;
using MoDi.Protocol;

namespace MoDi.Desktop.Core.Session;

public enum SessionControlAction : byte
{
    DisconnectRequest = 1,
    DisconnectAck = 2,
}

public enum DisconnectReason : byte
{
    UserSwitch = 1,
    Repair = 2,
    UserStop = 3,
}

public enum DisconnectResult : byte
{
    None = 0,
    Accepted = 1,
    Ignored = 2,
}

public readonly record struct SessionControlMessage(
    SessionControlAction Action,
    Guid SessionId,
    byte OldLink,
    byte TargetLink,
    DisconnectReason Reason,
    DisconnectResult Result)
{
    private const int MessageLength = 24;
    private const byte Domain = 0x0F;
    private const byte MessageType = 0x01;
    private const byte Version = 0x01;

    public static SessionControlMessage Request(
        Guid sessionId,
        byte oldLink,
        byte targetLink,
        DisconnectReason reason)
    {
        if (sessionId == Guid.Empty) throw new ArgumentOutOfRangeException(nameof(sessionId));
        if (!IsLinkType(oldLink)) throw new ArgumentOutOfRangeException(nameof(oldLink));
        if (!IsLinkType(targetLink)) throw new ArgumentOutOfRangeException(nameof(targetLink));
        return new(SessionControlAction.DisconnectRequest, sessionId, oldLink, targetLink, reason, DisconnectResult.None);
    }

    public static SessionControlMessage Ack(SessionControlMessage request, DisconnectResult result)
    {
        if (request.Action != SessionControlAction.DisconnectRequest)
            throw new ArgumentException("ACK requires a disconnect request.", nameof(request));
        if (result == DisconnectResult.None)
            throw new ArgumentOutOfRangeException(nameof(result), "ACK result must be accepted or ignored.");
        return request with { Action = SessionControlAction.DisconnectAck, Result = result };
    }

    public byte[] Encode()
    {
        var payload = new byte[MessageLength];
        payload[0] = Domain;
        payload[1] = MessageType;
        payload[2] = Version;
        payload[3] = (byte)Action;
        WriteGuid(SessionId, payload.AsSpan(4, 16));
        payload[20] = OldLink;
        payload[21] = TargetLink;
        payload[22] = (byte)Reason;
        payload[23] = (byte)Result;
        return payload;
    }

    public Packet ToPacket() => new()
    {
        Type = PacketType.Data,
        LinkType = OldLink,
        Sequence = 0,
        Payload = Encode(),
    };

    public static bool TryDecode(ReadOnlySpan<byte> payload, out SessionControlMessage message)
    {
        message = default;
        if (payload.Length != MessageLength || payload[0] != Domain || payload[1] != MessageType || payload[2] != Version)
            return false;
        if (!Enum.IsDefined((SessionControlAction)payload[3]) ||
            !IsLinkType(payload[20]) || !IsLinkType(payload[21]) ||
            !Enum.IsDefined((DisconnectReason)payload[22]) ||
            !Enum.IsDefined((DisconnectResult)payload[23]))
            return false;

        var action = (SessionControlAction)payload[3];
        var result = (DisconnectResult)payload[23];
        if (action == SessionControlAction.DisconnectRequest && result != DisconnectResult.None)
            return false;
        if (action == SessionControlAction.DisconnectAck && result == DisconnectResult.None)
            return false;

        message = new(
            action,
            ReadGuid(payload.Slice(4, 16)),
            payload[20],
            payload[21],
            (DisconnectReason)payload[22],
            result);
        return true;
    }

    private static bool IsLinkType(byte linkType) => linkType is
        LinkType.WifiLan or LinkType.WifiDirect or LinkType.Bluetooth or LinkType.Usb;

    internal static void WriteGuid(Guid value, Span<byte> destination) =>
        Convert.FromHexString(value.ToString("N")).CopyTo(destination);

    internal static Guid ReadGuid(ReadOnlySpan<byte> source) =>
        Guid.ParseExact(Convert.ToHexString(source), "N");
}

public readonly record struct HelloSessionIdentity(int Route, string? Token, Guid SessionId);

public readonly record struct HandshakeResult(int Route, Guid SessionId);

public static class HelloSessionPayload
{
    private const int TokenLength = 8;
    private const int SessionLength = 16;

    public static byte[] Encode(int route, string? token, Guid sessionId)
    {
        if (route is < 0 or > 3) throw new ArgumentOutOfRangeException(nameof(route));
        if (sessionId == Guid.Empty) throw new ArgumentOutOfRangeException(nameof(sessionId));
        byte[]? tokenBytes = token is null ? null : Encoding.ASCII.GetBytes(token);
        if (tokenBytes is { Length: < 1 or > TokenLength } ||
            token is not null && Encoding.ASCII.GetString(tokenBytes!) != token)
            throw new ArgumentException("Token must contain 1 to 8 ASCII bytes.", nameof(token));

        var sessionOffset = 1 + (tokenBytes is null ? 0 : TokenLength);
        var payload = new byte[sessionOffset + SessionLength];
        payload[0] = (byte)route;
        tokenBytes?.CopyTo(payload, 1);
        SessionControlMessage.WriteGuid(sessionId, payload.AsSpan(sessionOffset, SessionLength));
        return payload;
    }

    public static bool TryDecode(
        ReadOnlySpan<byte> payload,
        bool tokenRequired,
        out HelloSessionIdentity identity)
    {
        identity = default;
        var sessionOffset = tokenRequired ? 1 + TokenLength : 1;
        if (payload.Length != sessionOffset + SessionLength || payload[0] > 3)
            return false;

        string? token = null;
        if (tokenRequired)
        {
            var tokenBytes = payload.Slice(1, TokenLength);
            var terminator = tokenBytes.IndexOf((byte)0);
            var tokenLength = terminator < 0 ? TokenLength : terminator;
            if (tokenLength == 0 ||
                terminator >= 0 && tokenBytes.Slice(terminator).IndexOfAnyExcept((byte)0) >= 0 ||
                !AllPrintableAscii(tokenBytes.Slice(0, tokenLength)))
                return false;
            token = Encoding.ASCII.GetString(tokenBytes.Slice(0, tokenLength));
        }

        identity = new(
            payload[0],
            token,
            SessionControlMessage.ReadGuid(payload.Slice(sessionOffset, SessionLength)));
        return true;
    }

    public static bool MatchesAck(ReadOnlySpan<byte> payload, Guid expectedSessionId) =>
        TryDecode(payload, tokenRequired: false, out var identity) && identity.SessionId == expectedSessionId;

    private static bool AllPrintableAscii(ReadOnlySpan<byte> value)
    {
        foreach (var current in value)
            if (current is < 0x20 or > 0x7E)
                return false;
        return true;
    }
}
